sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuMultiFileReader.scala

```diff
@@ -618,14 +618,36 @@ abstract class MultiFileCloudPartitionReaderBase(
           // Filter time here includes the buffer time as well since those
           // happen in the same background threads. This is as close to wall
           // clock as we can get right now without further work.
-          val startTime = System.nanoTime()
-          val fileBufsAndMeta = getNextBuffersAndMeta()
-          val blockedTime = System.nanoTime() - startTime
-          metrics.get(FILTER_TIME).foreach {
-            _ += (blockedTime * fileBufsAndMeta.getFilterTimePct).toLong
-          }
-          metrics.get(BUFFER_TIME).foreach {
-            _ += (blockedTime * fileBufsAndMeta.getBufferTimePct).toLong
+          require(metrics.contains(BUFFER_TIME), "bufferTime should be set")
+          require(metrics.contains(FILTER_TIME), "filterTime should be set")
+
+          val fileBufsAndMeta = if (metrics.contains(BUFFER_TIME_WITH_SEM)) {
+            require(metrics.contains(FILTER_TIME_WITH_SEM),
+              "filterTimeWithSem should be set if bufferTimeWithSem is set")
+            val taskContext = TaskContext.get()
+            require(taskContext != null, "TaskContext should not be null")
+
+            val wallStart = metrics(BUFFER_TIME).value
+            val withSemStart = metrics(BUFFER_TIME_WITH_SEM).value
+            val ret = GpuMetric.withSemaphoreTime(metrics(BUFFER_TIME),
```

Comment:
nit: why pass in two temp metrics into withSemaphoreTime , and attribute  the temp metrics to related metrics afterwards? seems it would be simpler and clearer

Reply:
Good idea! Replace the placeholder with the a temp local GpuMetric.